### PR TITLE
Reference disk usage metric + refactoring [BT-20]

### PR DESCRIPTION
**Pull request by grsterin**

### CODE REVIEW

**Review by kshakir — COMMENTED** [review]
Main code LGTM so far. Due to parallel CI failures I can't tell if this was integration tested in centaur or not, but the changes do not appear to be unit tested by the `sbt` suite, especially `sendNumberOfReferenceFilesUsedGauge`.

Also, are there any docs on how to use this? For example, will this new metric automatically appear alongside the others in the metrics explorer on the [Cromwell Dashboard](https://console.cloud.google.com/monitoring/dashboards/builder/12440147339961148290?project=broad-dsde-prod), does something need to be done to further wire this in?

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala [hunk 1]
```diff
@@ -623,6 +637,20 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
     }
   }
 
+  private def sendNumberOfReferenceFilesUsedGauge(inputs: List[PipelinesApiInput]): Unit = {
+    (useReferenceDisks, jesAttributes.referenceFileToDiskImageMappingOpt) match {
+      case (true, Some(referenceFileToDiskImageMapping)) =>
+        val referenceInputsNumber = getReferenceInputsToMountedPathMappings(referenceFileToDiskImageMapping, inputs).size
+        sendGauge(
+          NonEmptyList.of("referencefiles", "used", "number"),
```
[the remark this answers is not in the document]
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala:645]: You mean incorporating file name into the metric "name"? I think the number of reference files is huge and in Grafana we'd need to manually add each metric to the dashboard to make it visible

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala [hunk 2]
```diff
@@ -623,6 +637,20 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
...
+          referenceInputsNumber.longValue,
+          InstrumentationPrefixes.JobPrefix
+        )
+      case _ =>
+        // reference disks feature is either not configured in Cromwell or disabled in workflow options
```
[the remark this answers is not in the document]
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala]: I'm not sure we currently write in the log if any of PAPI configuration attributes are specified or not.

**Review by kshakir — APPROVED** [review]
For folks that come across this PR in the future, we did NOT make time today to discuss where DSP metrics are supposed to be sent... so I believe this PR is still generating metrics aimed at the old dashboard and not what folks have been referring to as the "post-kubernetes" dashboard 🤷🦶🏈

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcovarr) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala:656]: Hmm I didn't think those needed to be created manually but I am not a Grafana expert. We certainly don't create new metrics manually when we spin up new k8s pods.

[the remark this answers is not in the document]
Comment (mcovarr) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala:656]: There are currently 30 reference files set to go out, 24 for HG 38 and 6 for HG 19.

File: supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/PipelinesApiAsyncBackendJobExecutionActor.scala [hunk 3]
```diff
@@ -176,52 +174,37 @@ class PipelinesApiAsyncBackendJobExecutionActor(standardParams: StandardAsyncExe
...
-      val referenceInputsToMountedPathsOpt: Option[Map[PipelinesApiInput, String]] =
-        referenceFileToDiskImageMappingOpt.map(getReferenceInputsToMountedPathMappings(_, inputs))
-
-      val referenceFilesLocalizationScript = {
-        val symlinkCreationCommandsOpt = referenceInputsToMountedPathsOpt map { referenceInputsToMountedPaths =>
-          referenceInputsToMountedPaths map {
-            case (input, absolutePathOnRefDisk) =>
-              s"mkdir -p ${input.containerPath.parent.pathAsString} && ln -s $absolutePathOnRefDisk ${input.containerPath.pathAsString}"
-          }
+    val referenceFilesLocalizationScript = {
```
Comment (mcovarr) [inline supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/PipelinesApiAsyncBackendJobExecutionActor.scala:181]: It seems surprising the lines below are marked as not being tested, unless Centaur coverage is not being included for some reason...

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (mcovarr) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala]: any chance of sneaking a one-liner like that into this PR or should we make a separate ticket?

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala:656]: Yeah, metrics have to be manually added to the dashboard to be visible. But now I remember that Grafana supports wildcards, so we can add metrics for all files as a single query

**Comment by grsterin** [comment]
Updated to publish counter metrics on per-reference-file basis (incremented until Cromwell restart). Also added logging for reference disks feature configuration step.

**Review by kshakir — APPROVED** [review]

File: supportedBackends/google/pipelines/common/src/test/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActorSpec.scala [hunk 4]
```diff
@@ -127,12 +131,13 @@ class PipelinesApiAsyncBackendJobExecutionActorSpec extends TestKitSuite
...
              jesSingletonActor: ActorRef = emptyActor,
-             ioActor: ActorRef = mockIoActor) = {
+             ioActor: ActorRef = mockIoActor,
+             serviceRegistryActor: ActorRef = kvService) = {
 
       this(
         DefaultStandardAsyncExecutionActorParams(
           jobIdKey = PipelinesApiAsyncBackendJobExecutionActor.JesOperationIdKey,
-          serviceRegistryActor = kvService,
+          serviceRegistryActor = serviceRegistryActor,
```
Comment (mcovarr) [inline supportedBackends/google/pipelines/common/src/test/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActorSpec.scala:140]: oops

**Review by mcovarr — APPROVED** [review]

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala [hunk 5]
```diff
@@ -623,6 +648,17 @@ class PipelinesApiAsyncBackendJobExecutionActor(override val standardParams: Sta
     }
   }
 
+  protected def sendIncrementMetricsForReferenceFiles(referenceInputFilesOpt: Option[Set[PipelinesApiInput]]): Unit = {
+    referenceInputFilesOpt match {
+      case Some(referenceInputFiles) =>
+        referenceInputFiles.foreach { referenceInputFile =>
+          increment(NonEmptyList.of("referencefiles", referenceInputFile.relativeHostPath.pathAsString))
```
[the remark this answers is not in the document]
Comment (grsterin) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiAsyncBackendJobExecutionActor.scala:655]: Yes, it's going to be the same as cloud path, but with `gs://` part dropped. Like `gcp-public-data--broad-references/hg19/v0/README`

**Pull request merged**